### PR TITLE
1168 feat single dataset outlier detection

### PR DESCRIPTION
**Pull request by Nadav-Barak**
Added a check that detects the ration of outliers in the dataset.
Under the hood uses Loop algorithm on top on a distance matrix computed based on Gower's method.

### CODE REVIEW

File: deepchecks/tabular/checks/integrity/outlier_detection.py [hunk 1]
```diff
@@ -0,0 +1,149 @@
...
+from deepchecks.core import CheckResult, ConditionResult, ConditionCategory
+from deepchecks.tabular import Context, SingleDatasetCheck
+from deepchecks.utils import gower_distance
+from deepchecks.utils.dataframes import select_from_dataframe
+from deepchecks.utils.typing import Hashable
+
+__all__ = ['OutlierDetection']
+
+
+class OutlierDetection(SingleDatasetCheck):
```
Comment (matanper) [inline deepchecks/tabular/checks/integrity/outlier_detection.py:27]: The name is a bit ambigious. If we go by the naming of the drift checks, it should be something like "WholeDatasetOutliers"

File: deepchecks/tabular/checks/integrity/outlier_detection.py [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+    ignore_columns : Union[Hashable, List[Hashable]] , default: None
+        Columns to ignore, if none given checks based on columns variable
+    num_nearest_neighbors : int, default: 10
+        Number of nearest neighbors to use for outlier detection.
+    extend_parameter: int, default: 3
+        Extend parameter for LoOP algorithm.
+    n_samples : int , default: 50_000
+        number of samples to use for this check.
+    n_to_show : int , default: 5
+        number of most data elements with the highest outlier score to show.
```
Comment (matanper) [inline deepchecks/tabular/checks/integrity/outlier_detection.py:43]: "number of most data elements with..." valid phrasing?

File: deepchecks/tabular/checks/integrity/outlier_detection.py [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+        """Run check."""
+        if dataset_type == 'train':
+            dataset = context.train
+        else:
+            dataset = context.test
+        dataset = dataset.sample(self.n_samples, random_state=self.random_state, drop_na_label=True)
+        df = select_from_dataframe(dataset.data, self.columns, self.ignore_columns)
+
+        # Calculate distances matrix and retrieve nearest neighbors based on distance matrix.
+        cat_features = df.columns.map(lambda x: x in dataset.cat_features)
```
Comment (matanper) [inline deepchecks/tabular/checks/integrity/outlier_detection.py:82]: isn't the result of this is boolean array (since it's `map` not `filter`)? if so, it's not so clear

File: deepchecks/tabular/checks/integrity/outlier_detection.py [hunk 4]
```diff
@@ -0,0 +1,149 @@
...
+        nn_matrix = knn_distance_based.kneighbors(return_distance=False)
+
+        # Calculate outlier probability score using loop algorithm.
+        m = loop.LocalOutlierProbability(distance_matrix=dist_matrix, neighbor_matrix=nn_matrix,
+                                         extent=self.extend_parameter, n_neighbors=self.num_nearest_neighbors).fit()
+        prob_vector = m.local_outlier_probabilities
+
+        # Create the check result visualization
+        top_n_idx = np.argsort(prob_vector)[-self.n_to_show:]
+        top_n_values = [prob_vector[i] for i in top_n_idx]
```
Comment (matanper) [inline deepchecks/tabular/checks/integrity/outlier_detection.py:95]: I think `prob_vector[top_n_idx]` will also work

File: deepchecks/tabular/checks/integrity/outlier_detection.py [hunk 5]
```diff
@@ -0,0 +1,149 @@
...
+        headnote = """<span>
+                    The Outlier Probability Score is calculated by the LoOP algorithm which measures the local deviation
+                     of density of a given sample with respect to its neighbors. These outlier scores are directly
+                      interpretable as a probability of an object being an outlier. see 
+                      <a href="https://www.dbs.ifi.lmu.de/Publikationen/Papers/LoOP1649.pdf" target="_blank" rel="noopener noreferrer">link</a>
+                       for more information.<br><br>
+                </span>"""
+
+        quantiles_vector = np.quantile(prob_vector, np.array(range(1000)) / 1000, method='closest_observation')
+        return CheckResult([quantiles_vector, df.shape[0]], display=[headnote, dataset_outliers])
```
Comment (matanper) [inline deepchecks/tabular/checks/integrity/outlier_detection.py:109]: what `df.shape[0]` represents? better return dict with names so it will be clearer

File: deepchecks/tabular/checks/integrity/outlier_detection.py [hunk 6]
```diff
@@ -0,0 +1,149 @@
...
+        Parameters
+        ----------
+        outlier_score_threshold : float, default: 0.8
+            Outlier probability score threshold to be considered outlier.
+        """
+        name = f'No outliers in dataset over score threshold of {outlier_score_threshold}'
+        return self.add_condition(name, _condition_outliers_number, outlier_score_threshold=outlier_score_threshold)
+
+
+def _condition_outliers_number(result: np.ndarray, outlier_score_threshold: float,
```
Comment (matanper) [inline deepchecks/tabular/checks/integrity/outlier_detection.py:139]: the result is python list

File: deepchecks/tabular/checks/integrity/outlier_detection.py (hunk 3 above)
Comment (matanper) [reply deepchecks/tabular/checks/integrity/outlier_detection.py:82]: Also we have more types than categorical and numerical, so if I understand correctly you want `x not in dataset.numerical_features`

File: tests/checks/integrity/outlier_detection_test.py [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+
+
+def test_integer_single_column_no_nulls():
+    # Arrange
+    data = {'col1': [1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1000]}
+    dataframe = pd.DataFrame(data=data)
+    # Act
+    result = OutlierDetection(n_to_show=1, num_nearest_neighbors=5).run(dataframe)
+    # Assert
+    assert_that(max(result.value[0]) > 0.7)
```
Comment (matanper) [inline tests/checks/integrity/outlier_detection_test.py:35]: the "hamcrest way" to do that would be
`assert_that(result.value[0], has_item(greater_than(0.7))`

File: tests/checks/integrity/outlier_detection_test.py [hunk 8]
```diff
@@ -0,0 +1,99 @@
...
+
+
+def test_iris_regular():
+    # Arrange
+    iris = load_iris()
+    # Act
+    result = OutlierDetection(n_to_show=1, num_nearest_neighbors=5).run(pd.DataFrame(iris.data))
+    # Assert
+    print(max(result.value[0]))
```
Comment (matanper) [inline tests/checks/integrity/outlier_detection_test.py:87]: remove print

File: tests/checks/integrity/outlier_detection_test.py [hunk 9]
```diff
@@ -0,0 +1,99 @@
...
+    data = {'col1': ['a', 'a', 'a', 'a', 'a', pd.NA, 'a', 'a', 'a', 'a', np.nan, 'a', 'a', 'b'],
+            'col2': [1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1000]}
+    dataset = Dataset(pd.DataFrame(data=data), cat_features=['col1'])
+    # Act
+    result = OutlierDetection(n_to_show=1, num_nearest_neighbors=5).run(dataset)
+    # Assert
+    assert_that(max(result.value[0]) > 0.7)
+
+
+def test_iris_regular():
```
Comment (matanper) [inline tests/checks/integrity/outlier_detection_test.py:81]: you should load iris as a fixture (parameter in the test def), instead of loading it directly

**Review by matanper — COMMENTED** [review]
Overall looking very good for a first check. Made small comments

File: deepchecks/utils/gower_distance.py [hunk 10]
```diff
@@ -0,0 +1,96 @@
...
+    cat_features: numpy.array
+        Boolean array of representing which of the columns are categorical features.
+
+    Returns numpy.ndarray representing the distance matrix.
+    """
+    if not isinstance(data, np.ndarray): data = np.asarray(data)
+    ranges = calculate_ranges(data, cat_features)
+
+    result = np.zeros((data.shape[0], data.shape[0]))
+    for i in range(data.shape[0]):
```
[the remark this answers is not in the document]
Comment (noamzbr) [reply deepchecks/utils/gower_distance.py:36]: Or other forms of vectorization

File: deepchecks/tabular/checks/integrity/outlier_detection.py [hunk 11]
```diff
@@ -0,0 +1,149 @@
...
+from deepchecks.tabular import Context, SingleDatasetCheck
+from deepchecks.utils import gower_distance
+from deepchecks.utils.dataframes import select_from_dataframe
+from deepchecks.utils.typing import Hashable
+
+__all__ = ['OutlierDetection']
+
+
+class OutlierDetection(SingleDatasetCheck):
+    """Detects outliers in a dataset using the LoOP algorithm.
```
Comment (noamzbr) [inline deepchecks/tabular/checks/integrity/outlier_detection.py:28]: Below this "summery line" there should be a more elaborate description explaining that this is "Multivariate" outlier detection (considers multiple features at once) and a bit about the LoOP algo (also link to paper).

File: deepchecks/tabular/checks/integrity/outlier_detection.py [hunk 12]
```diff
@@ -0,0 +1,149 @@
...
+
+        headnote = """<span>
+                    The Outlier Probability Score is calculated by the LoOP algorithm which measures the local deviation
+                     of density of a given sample with respect to its neighbors. These outlier scores are directly
+                      interpretable as a probability of an object being an outlier. see 
+                      <a href="https://www.dbs.ifi.lmu.de/Publikationen/Papers/LoOP1649.pdf" target="_blank" rel="noopener noreferrer">link</a>
+                       for more information.<br><br>
+                </span>"""
+
+        quantiles_vector = np.quantile(prob_vector, np.array(range(1000)) / 1000, method='closest_observation')
```
Comment (noamzbr) [inline deepchecks/tabular/checks/integrity/outlier_detection.py:108]:
```suggestion
        quantiles_vector = np.quantile(prob_vector, np.array(range(0, 1, 0.001)), method='closest_observation')
```

File: deepchecks/tabular/checks/integrity/outlier_detection.py [hunk 13]
```diff
@@ -0,0 +1,149 @@
...
+        """Add condition - no more than given ratio of elements over outlier threshold are allowed.
+        Parameters
+        ----------
+        max_outliers_ratio : float , default: 0.005
+            Maximum ratio of outliers allowed in dataset.
+        outlier_score_threshold : float, default: 0.8
+            Outlier probability score threshold to be considered outlier.
+        """
+        if max_outliers_ratio > 1 or max_outliers_ratio < 0:
+            raise ValueError('max_outliers_ratio must be between 0 and 1')
```
Comment (noamzbr) [inline deepchecks/tabular/checks/integrity/outlier_detection.py:122]:
```suggestion
            raise DeepchecksValueError('max_outliers_ratio must be between 0 and 1')
```
This makes sure it won't kill the suite it's running inside.

File: deepchecks/tabular/checks/integrity/outlier_detection.py [hunk 14]
```diff
@@ -0,0 +1,149 @@
...
+
+
+def _condition_outliers_number(result: np.ndarray, outlier_score_threshold: float,
+                               max_outliers_ratio: float = 0):
+    max_outliers_ratio = round(max_outliers_ratio, 3)
+    quantiles_vector = result[0]
+
+    if quantiles_vector[int(1000 - max_outliers_ratio * 1000)] > outlier_score_threshold:
+        ratio_above_threshold = round((1000 - np.argmax(quantiles_vector > outlier_score_threshold)) / 10)
+        details = f'{ratio_above_threshold}% of dataset above outlier threshold of {outlier_score_threshold}'
```
Comment (noamzbr) [inline deepchecks/tabular/checks/integrity/outlier_detection.py:146]: Threshold is in name so no need to write it again here.

File: deepchecks/tabular/checks/integrity/outlier_detection.py [hunk 15]
```diff
@@ -0,0 +1,149 @@
...
+        Parameters
+        ----------
+        max_outliers_ratio : float , default: 0.005
+            Maximum ratio of outliers allowed in dataset.
+        outlier_score_threshold : float, default: 0.8
+            Outlier probability score threshold to be considered outlier.
+        """
+        if max_outliers_ratio > 1 or max_outliers_ratio < 0:
+            raise ValueError('max_outliers_ratio must be between 0 and 1')
+        name = f'Not more than {max_outliers_ratio * 100}% of dataset over outlier score of {outlier_score_threshold}'
```
Comment (noamzbr) [inline deepchecks/tabular/checks/integrity/outlier_detection.py:123]: When printing numbers or percents, use format_percent and format_number

File: deepchecks/tabular/checks/integrity/outlier_detection.py (hunk 14 above)
Comment (noamzbr) [reply deepchecks/tabular/checks/integrity/outlier_detection.py:146]: Same here, format_percent

Comment (noamzbr) [inline deepchecks/tabular/checks/integrity/outlier_detection.py:146]:
```suggestion
        details = f'{ratio_above_threshold}% of dataset samples above outlier threshold of {outlier_score_threshold}'
```

File: deepchecks/tabular/checks/integrity/outlier_detection.py [hunk 16]
```diff
@@ -0,0 +1,149 @@
...
+
+    def add_condition_no_outliers(self, outlier_score_threshold: float = 0.8):
+        """Add condition - no elements over outlier threshold are allowed.
+
+        Parameters
+        ----------
+        outlier_score_threshold : float, default: 0.8
+            Outlier probability score threshold to be considered outlier.
+        """
+        name = f'No outliers in dataset over score threshold of {outlier_score_threshold}'
```
Comment (noamzbr) [inline deepchecks/tabular/checks/integrity/outlier_detection.py:135]:
```suggestion
        name = f'No samples in dataset over outlier score of {outlier_score_threshold}'
```

File: deepchecks/tabular/checks/integrity/outlier_detection.py [hunk 17]
```diff
@@ -0,0 +1,149 @@
...
+                     of density of a given sample with respect to its neighbors. These outlier scores are directly
+                      interpretable as a probability of an object being an outlier. see 
+                      <a href="https://www.dbs.ifi.lmu.de/Publikationen/Papers/LoOP1649.pdf" target="_blank" rel="noopener noreferrer">link</a>
+                       for more information.<br><br>
+                </span>"""
+
+        quantiles_vector = np.quantile(prob_vector, np.array(range(1000)) / 1000, method='closest_observation')
+        return CheckResult([quantiles_vector, df.shape[0]], display=[headnote, dataset_outliers])
+
+    def add_condition_not_more_outliers_than(self, max_outliers_ratio: float = 0.005,
```
Comment (noamzbr) [inline deepchecks/tabular/checks/integrity/outlier_detection.py:111]:
```suggestion
    def add_condition_outlier_ratio_not_greater_than(self, max_outliers_ratio: float = 0.005,
```

File: deepchecks/utils/gower_distance.py [hunk 18]
```diff
@@ -0,0 +1,96 @@
...
+import numpy as np
+import pandas as pd
+
+
+def gower_matrix(data: np.ndarray, cat_features: np.array) -> np.ndarray:
+    """
+    Calculate distance matrix for a dataset using Gower's method.
+
+    Can deal with missing values.
+    Requires each two samples to have at least one sheared non-null value.
```
[the remark this answers is not in the document]
Comment (Nadav-Barak) [reply deepchecks/utils/gower_distance.py:22]: This comment speaks about a more unique case where we have only numerical features and two examples that do not have a single jointly non null element, rephrased in code.

[the remark this answers is not in the document]
Comment (Nadav-Barak) [reply deepchecks/utils/gower_distance.py:22]: In second thought removed, this is a really rare case.

File: deepchecks/utils/gower_distance.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (Nadav-Barak) [reply deepchecks/utils/gower_distance.py:36]: from vectorization: The [vectorize] function is provided primarily for convenience, not for performance. The implementation is essentially a for loop.
It will be less efficient in a vectorize manner due to double calculations for result[i, j], result[j, i]

File: deepchecks/utils/gower_distance.py [hunk 19]
```diff
@@ -0,0 +1,96 @@
...
+    Parameters
+    ----------
+    data: numpy.ndarray
+        Dataset matrix.
+    cat_features: numpy.array
+        Boolean array of representing which of the columns are categorical features.
+
+    Returns numpy.array representing the ranges of each numeric feature and -1 for categorical.
+    """
+    ranges = np.zeros(data.shape[1])
```
[the remark this answers is not in the document]
Comment (Nadav-Barak) [reply deepchecks/utils/gower_distance.py:57]: Great! replaced

File: deepchecks/utils/gower_distance.py [hunk 20]
```diff
@@ -0,0 +1,96 @@
...
+        else:
+            # numeric feature
+            if pd.isnull(vec1[col_index]) or pd.isnull(vec2[col_index]):
+                continue
+            sum_dist += np.abs(vec1[col_index] - vec2[col_index]) / ranges[col_index]
+            num_features += 1
+
+    if num_features == 0:
+        raise ValueError(f"No non null features found to compare examples {vec1} and {vec2}.")
+    return sum_dist / num_features
```
[the remark this answers is not in the document]
Comment (Nadav-Barak) [reply deepchecks/utils/gower_distance.py:96]: It is mathematically equivalent

File: deepchecks/tabular/checks/integrity/outlier_detection.py (hunk 12 above)
Comment (Nadav-Barak) [reply deepchecks/tabular/checks/integrity/outlier_detection.py:108]: Doesnt work, range only accepts integers

File: deepchecks/utils/gower_distance.py [hunk 21]
```diff
@@ -0,0 +1,96 @@
...
+    ----------
+        vec1: First vector.
+        vec2: Second vector.
+        ranges: Ranges of each numeric feature or -1 for categorical.
+
+    Returns float representing Gower's distance between two vectors.
+    """
+    sum_dist = 0
+    num_features = 0
+    for col_index in range(len(vec1)):
```
[the remark this answers is not in the document]
Comment (Nadav-Barak) [reply deepchecks/utils/gower_distance.py:79]: I disagree, it contains many different options and will be hard to follow logic in a one/two liners

File: deepchecks/tabular/checks/integrity/outlier_detection.py (hunk 1 above)
Comment (noamzbr) [reply deepchecks/tabular/checks/integrity/outlier_detection.py:28]: Technically correct, but I must admit I've grown to dislike the "Whole" thing....
Maybe OutlierSampleDetection?

File: deepchecks/tabular/checks/integrity/outlier_detection.py [hunk 22]
```diff
@@ -0,0 +1,157 @@
...
+            dataset = context.test
+        dataset = dataset.sample(self.n_samples, random_state=self.random_state, drop_na_label=True)
+        df = select_from_dataframe(dataset.data, self.columns, self.ignore_columns)
+
+        # Calculate distances matrix and retrieve nearest neighbors based on distance matrix.
+        df_cols_for_gower = df[dataset.cat_features + dataset.numerical_features]
+        is_categorical_arr = df_cols_for_gower.columns.map(lambda x: x in dataset.cat_features)
+        dist_matrix = gower_distance.gower_matrix(np.asarray(df_cols_for_gower),
+                                                  cat_features=np.array(is_categorical_arr, dtype=bool))
+        dist_matrix[np.isnan(dist_matrix)] = np.nanmean(dist_matrix)
```
Comment (noamzbr) [inline deepchecks/tabular/checks/integrity/outlier_detection.py:92]: I know it's a remote possibility, but let's catch the case when all are nan (think this can happen if the sample is all nulls) and just ignore that sample?

File: deepchecks/utils/gower_distance.py (hunk 20 above)
[the remark this answers is not in the document]
Comment (noamzbr) [reply deepchecks/utils/gower_distance.py:88]: Sure, I didn't follow the mathematics all the way, just wanted to make sure.

File: deepchecks/utils/gower_distance.py (hunk 21 above)
[the remark this answers is not in the document]
Comment (noamzbr) [reply deepchecks/utils/gower_distance.py:71]: why? We can do it pretty simply IMO
```
vec1_cat = ..
vec2_cat = ..
vec1_num = ..
vec2_num = ..
cat_1_nan = pd.isnull(vec1_cat)
cat_2_nan = pd.isnull(vec2_cat)
n_both_cat_are_null = (cat_1_nan & cat_1_nan).sum()
if n_both_cat_are_null == len(vec1):
   return np.nan
sum_dist = 0
sum_dist  += (cat_1_nan  ^ cat_2_nan ).sum()  # Add xor of nulls
any_nulls = cat_1_nan  | cat_2_nan 
sum_dist  += (vec1_cat[~any_nulls] !=  vec2_cat[~any_nulls])
num_nans = pd.isnull(vec1_num) | pd.isnull(vec2_num)
sum_dist  += (vec1_num[~num_nans] - vec2_num[~num_nans]).abs / range_per_feature

return sum_dist / (len(vec1) - n_both_cat_are_null - len(num_nans ))
```

File: deepchecks/tabular/checks/integrity/outlier_detection.py (hunk 1 above)
Comment (Nadav-Barak) [reply deepchecks/tabular/checks/integrity/outlier_detection.py:28]: Works for me :)

**Review by ItayGabbay — COMMENTED** [review]
Really nice PR! 
2 points:
1. I think it's worth having a separate unit test module for the gower_distance module. It seems like it has many edge cases and we should cover them all.
2. Missing example.

Overall the code is really understandable and clear, great work! (Although I wish we can get rid of the `PyNomaly` requirement, (I know it's impossible to do it :) ))

File: deepchecks/tabular/checks/integrity/outlier_sample_detection.py [hunk 23]
```diff
@@ -0,0 +1,160 @@
...
+logger = logging.getLogger('deepchecks')
+
+
+class OutlierSampleDetection(SingleDatasetCheck):
+    """Detects outliers in a dataset using the LoOP algorithm.
+
+    The LoOP algorithm is a robust method for detecting outliers in a dataset across multiple variables by comparing
+    the density in the area of a sample with the densities in the areas of its nearest neighbors.
+    See https://www.dbs.ifi.lmu.de/Publikationen/Papers/LoOP1649.pdf for more details.
+    LoOP is build upon a distance matrix, in our implementation we use the Gower distance that measure the distance
```
Comment (noamzbr) [inline deepchecks/tabular/checks/integrity/outlier_sample_detection.py:36]:
```suggestion
    LoOP relies on a distance matrix. in our implementation we use the Gower distance that measure the distance
```

File: deepchecks/tabular/checks/integrity/outlier_sample_detection.py [hunk 24]
```diff
@@ -0,0 +1,160 @@
...
+        is_categorical_arr = np.array(df_cols_for_gower.columns.map(lambda x: x in dataset.cat_features), dtype=bool)
+        dist_matrix, idx_matrix = gower_distance.gower_matrix_n_closets(data=np.asarray(df_cols_for_gower),
+                                                                        cat_features=is_categorical_arr,
+                                                                        num_neighbours=self.num_nearest_neighbors)
+
+        # Calculate outlier probability score using loop algorithm.
+        m = loop.LocalOutlierProbability(distance_matrix=dist_matrix, neighbor_matrix=idx_matrix,
+                                         extent=self.extent_parameter, n_neighbors=self.num_nearest_neighbors).fit()
+        prob_vector = np.asarray(m.local_outlier_probabilities, dtype=float)
+        prob_vector[np.isnan(prob_vector)] = 0
```
Comment (noamzbr) [inline deepchecks/tabular/checks/integrity/outlier_sample_detection.py:103]: Add a comment here about our decision to assign 0 outlier proba to samples that are all nans.

File: deepchecks/tabular/checks/integrity/outlier_sample_detection.py [hunk 25]
```diff
@@ -0,0 +1,160 @@
...
+        ----------
+        max_outliers_ratio : float , default: 0.005
+            Maximum ratio of outliers allowed in dataset.
+        outlier_score_threshold : float, default: 0.8
+            Outlier probability score threshold to be considered outlier.
+        """
+        if max_outliers_ratio > 1 or max_outliers_ratio < 0:
+            raise DeepchecksValueError('max_outliers_ratio must be between 0 and 1')
+        name = f'Not more than {format_percent(max_outliers_ratio)} of dataset over ' \
+               f'outlier score {outlier_score_threshold} '
```
Comment (noamzbr) [inline deepchecks/tabular/checks/integrity/outlier_sample_detection.py:135]:
```suggestion
               f'outlier score {format_number(outlier_score_threshold)} '
```

File: deepchecks/tabular/checks/integrity/outlier_sample_detection.py [hunk 26]
```diff
@@ -0,0 +1,160 @@
...
+
+    def add_condition_no_outliers(self, outlier_score_threshold: float = 0.8):
+        """Add condition - no elements over outlier threshold are allowed.
+
+        Parameters
+        ----------
+        outlier_score_threshold : float, default: 0.8
+            Outlier probability score threshold to be considered outlier.
+        """
+        name = f'No samples in dataset over outlier score of {outlier_score_threshold}'
```
Comment (noamzbr) [inline deepchecks/tabular/checks/integrity/outlier_sample_detection.py:147]:
```suggestion
        name = f'No samples in dataset over outlier score of {format_number(outlier_score_threshold)}'
```

File: deepchecks/utils/gower_distance.py [hunk 27]
```diff
@@ -0,0 +1,138 @@
...
+
+    return result
+
+
+def gower_matrix_n_closets(data: np.ndarray, cat_features: np.array, num_neighbours: int):
+    """
+    Calculate distance matrix for a dataset using Gower's method.
+
+    Can deal with missing values.
+    See https://statisticaloddsandends.wordpress.com/2021/02/23/what-is-gowers-distance/ for further details.
```
Comment (noamzbr) [inline deepchecks/utils/gower_distance.py:60]: If this is the "entry" function for gower this needs to be a bit fuller (and mainly explain the "trick" here of saving only the num_neighbours per sample.

File: deepchecks/tabular/checks/integrity/outlier_sample_detection.py [hunk 28]
```diff
@@ -0,0 +1,160 @@
...
+        dataset = dataset.sample(self.n_samples, random_state=self.random_state, drop_na_label=True)
+        df = select_from_dataframe(dataset.data, self.columns, self.ignore_columns)
+        if self.num_nearest_neighbors >= len(df):
+            logger.warning('Passed num_nearest_neighbors is greater than the number of samples in the dataset')
+            self.num_nearest_neighbors = len(df) - 1
+
+        # Calculate distances matrix and retrieve nearest neighbors based on distance matrix.
+        df_cols_for_gower = df[dataset.cat_features + dataset.numerical_features]
+        is_categorical_arr = np.array(df_cols_for_gower.columns.map(lambda x: x in dataset.cat_features), dtype=bool)
+        dist_matrix, idx_matrix = gower_distance.gower_matrix_n_closets(data=np.asarray(df_cols_for_gower),
```
Comment (noamzbr) [inline deepchecks/tabular/checks/integrity/outlier_sample_detection.py:95]: Bonus points - catching the out of memory exception that may raise if the user by mistake sets num_nearest_neighbors to be same order of magnitude as len(df) and raise instead a deepchecks error telling him to select a lower value.

File: docs/source/examples/tabular/checks/integrity/source/plot_outlier_sample_detection.py [hunk 29]
```diff
@@ -0,0 +1,68 @@
...
+**Structure:**
+
+* `How deepchecks detects outliers <#How-deepchecks-detects-outliers>`__
+* `Prepare data <#prepare-data>`__
+* `Run the check <#run-the-check>`__
+* `Define a condition <#define-a-condition>`__
+
+How deepchecks detects outliers
+========================
+Outlier Sample Detection searches for outliers samples using the LoOP algorithm.
```
Comment (noamzbr) [inline docs/source/examples/tabular/checks/integrity/source/plot_outlier_sample_detection.py:17]: Maybe first a word on the fact that the outliers are multivariate - considering all features at once.

File: docs/source/examples/tabular/checks/integrity/source/plot_outlier_sample_detection.py [hunk 30]
```diff
@@ -0,0 +1,68 @@
...
+The LoOP algorithm is a robust method for detecting outliers in a dataset across multiple variables by comparing
+the density in the area of a sample with the densities in the areas of its nearest neighbors
+(see `link <https://www.dbs.ifi.lmu.de/Publikationen/Papers/LoOP1649.pdf>`_ for further details).
+
+LoOP is build upon a distance matrix, in our implementation we use the Gower distance that averages the distances
```
Comment (noamzbr) [inline docs/source/examples/tabular/checks/integrity/source/plot_outlier_sample_detection.py:22]:
```suggestion
LoOP relies upon a distance matrix. In our implementation we use the Gower distance that averages the distances
```

File: docs/source/examples/tabular/checks/integrity/source/plot_outlier_sample_detection.py [hunk 31]
```diff
@@ -0,0 +1,68 @@
...
+per feature between samples. For numeric features it calculates the absolute distance divide by the range of the
```
Comment (noamzbr) [inline docs/source/examples/tabular/checks/integrity/source/plot_outlier_sample_detection.py:23]:
```suggestion
per feature between samples. For numeric features it calculates the absolute distance divided by the range of the
```

**Review by noamzbr — COMMENTED** [review]
Nice touch with the `gower_matrix_n_closets`! Could you please attach the html / pdf / screenshot of the example?

**Review by noamzbr — APPROVED** [review]
image

**Review by ItayGabbay — APPROVED** [review]
